firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java

```diff
@@ -834,6 +965,19 @@ private void applySetterAnnotations(Method method) {
                 + " is annotated with @ServerTimestamp but should not be. @ServerTimestamp can"
                 + " only be applied to fields and getters, not setters.");
       }
+
+      if (method.isAnnotationPresent(DocumentId.class)) {
+        Class<?> paramType = method.getParameterTypes()[0];
+        if (paramType != String.class && paramType != DocumentReference.class) {
+          throw new IllegalArgumentException(
+              "Method "
+                  + method.getName()
+                  + " is annotated with @DocumentId but sets "
+                  + paramType
+                  + " instead of String or DocumentReference.");
+        }
```

Comment:
Consider adding a helper to avoid replicating these checks / error message.

```
  private void validateDocumentIdType(String fieldDescription, String operation, Type type) {
      if (type != String.class && type != DocumentReference.class) {
         throw new IllegalArgumentException(
            fieldDescription
                + " is annotated with @DocumentId but "
                + operation
                + type
                + " instead of String or DocumentReference.");
        }
  }
```

Reply:
Done.